src/Lean/Meta/Match/Match.lean

```diff
@@ -804,22 +804,25 @@ def mkMatcher (input : MkMatcherInput) : MetaM MatcherResult :=
       unusedAltIdxs := unusedAltIdxs.reverse,
       addMatcher }
 
-def withMkMatcherInput (matcherName : Name) (k : MkMatcherInput → MetaM α) : MetaM α := do
+def getMkMatcherInputInContext (matcherApp : MatcherApp) : MetaM MkMatcherInput := do
+  let matcherName := matcherApp.matcherName
   let some matcherInfo ← getMatcherInfo? matcherName | throwError "not a matcher: {matcherName}"
   let matcherConst ← getConstInfo matcherName
-  forallBoundedTelescope matcherConst.type (some matcherInfo.arity) fun xs t => do
-  let params := xs[:matcherInfo.numParams]
-  let motive := xs[matcherInfo.numParams]
-  let discrs := xs[matcherInfo.numParams + 1:matcherInfo.numParams + 1 + matcherInfo.numDiscrs]
-  let alts := xs[matcherInfo.numParams + 1 + matcherInfo.numDiscrs:]
-  let u :=
-    if let some idx := matcherInfo.uElimPos?
-    then mkLevelParam matcherConst.levelParams.toArray[idx]
-    else levelZero
-  let matchType ← mkForallFVars discrs (mkConst ``PUnit [u])
-  let lhss ← alts.toArray.mapIdxM fun idx t => do
-    let ty ← inferType t
-    forallTelescope ty fun xs body => do
+  let matcherType ← instantiateForall matcherConst.type $ matcherApp.params ++ #[matcherApp.motive] 
+  let matchType ← do
+    let u := 
+      if let some idx := matcherInfo.uElimPos? 
+      then mkLevelParam matcherConst.levelParams.toArray[idx]
+      else levelZero
+    
+    forallBoundedTelescope matcherType (some matcherInfo.numDiscrs) fun discrs t => do
+    mkForallFVars discrs (mkConst ``PUnit [u])
+
+  let matcherType ← instantiateForall matcherType matcherApp.discrs 
+  let lhss ← Array.toList $ ←forallBoundedTelescope matcherType (some matcherApp.alts.size) fun alts _ => 
```

Comment:
`Array.toList` is pure. We should write `let lhss := Array.toList ...`
The not-yet-implemented linter should generate a warning in the future.